File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/sorting/Problem2473.java
package com.keunwon.algorithm.sorting;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Problem2473 {

    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        final int N = Integer.parseInt(br.readLine());
        long[] arr = new long[N];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Long.parseLong(st.nextToken());
        }

        Arrays.sort(arr);

        long max = Long.MAX_VALUE;
        long[] result = new long[3];

        for (int i = 0; i < N; i++) {
            int left = i + 1;
            int right = N - 1;

            while (left < right) {
                final long sum = arr[i] + arr[left] + arr[right];
                final long abaSum = Math.abs(sum);

                if (abaSum < max) {
                    max = abaSum;
                    result[0] = arr[i];
                    result[1] = arr[left];
                    result[2] = arr[right];
                }

                if (sum > 0) {
                    right--;
                } else {
                    left++;
                }
            }
        }

        for (long l : result) {
            System.out.print(l + " ");
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem2473.kt
package com.keunwon.algorithm.baekjoon

import kotlin.math.absoluteValue

/**
 * Title: 세 용액
 * Level: 골드-3
 **/
class Problem2473 {
    fun solution(arr: LongArray): LongArray {
        arr.sort()

        val picks = LongArray(3)
        var gap = Long.MAX_VALUE

        for (i in arr.indices) {
            var left = i
            var mid = i + 1
            var right = arr.lastIndex

            while (mid < right) {
                val sum = arr[left] + arr[mid] + arr[right]

                if (gap > sum.absoluteValue) {
                    gap = sum.absoluteValue
                    picks[0] = arr[left]
                    picks[1] = arr[mid]
                    picks[2] = arr[right]
                }

                when {
                    sum == 0L -> break
                    sum < 0L -> mid++
                    sum > 0L -> right--
                }
            }
        }
        return picks.sortedArray()
    }
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val n = br.readLine().toInt()
        val arr = br.readLine().split(" ").map { it.toLong() }.toLongArray()

        Problem2473().solution(arr).forEach { bw.write("$it ") }
        bw.flush()
    }
}
